pkg/recommendation/factory.go

```diff
@@ -0,0 +1,72 @@
+package recommendation
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/kubeshop/botkube/pkg/config"
+	"github.com/kubeshop/botkube/pkg/events"
+)
+
+// Recommendation performs checks for a given event.
+type Recommendation interface {
+	Do(ctx context.Context, event events.Event) (Result, error)
+	Name() string
+}
+
+// Result is the result of a recommendation check.
+type Result struct {
+	Info     []string
+	Warnings []string
+}
+
+// Factory is a factory for creating recommendation sets.
+type Factory struct {
+	logger     logrus.FieldLogger
+	dynamicCli dynamic.Interface
+}
+
+// NewFactory creates a new Factory instance.
+func NewFactory(logger logrus.FieldLogger, dynamicCli dynamic.Interface) *Factory {
+	return &Factory{logger: logger, dynamicCli: dynamicCli}
+}
+
+// NewForSources merges recommendation options from multiple sources, and creates a new Set.
+func (f *Factory) NewForSources(sources config.IndexableMap[config.Sources]) Set {
+	set := make(map[string]Recommendation)
+	for _, source := range sources {
+		f.createIfNotExist(source.Recommendations.Pod.LabelsSet, set,
+			func() Recommendation { return NewPodLabelsSet() },
+		)
+		f.createIfNotExist(source.Recommendations.Pod.NoLatestImageTag, set,
+			func() Recommendation { return NewPodNoLatestImageTag() },
+		)
+		f.createIfNotExist(source.Recommendations.Ingress.BackendServiceValid, set,
+			func() Recommendation { return NewIngressBackendServiceValid(f.dynamicCli) },
+		)
+		f.createIfNotExist(source.Recommendations.Ingress.TLSSecretValid, set,
+			func() Recommendation { return NewIngressTLSSecretValid(f.dynamicCli) },
+		)
+	}
+	return newRecommendationsSet(f.logger, set)
+}
+
+func (f *Factory) createIfNotExist(condition bool, set map[string]Recommendation, constructorFn func() Recommendation) {
+	if !condition {
+		return
+	}
+
+	// this solution has a downside that we need to create an instance every time,
```

Comment:
**nit**, you can just use your own `tag` for `createIfNotExist` func:

```go
const ingressTLSRec = "ingres-tls-rec"
		f.createIfNotExist(ingressTLSRec, source.Recommendations.Ingress.TLSSecretValid, set,
			func() Recommendation { return NewIngressTLSSecretValid(f.dynamicCli) },
		)
```
or even
```go
const (
	IngressTLS Rec = iota
	PodLabels
	// ....    
)
```

Reply:
If you don't mind I would keep it as it is as for me is less error-prone aporoach from the extensibility point of view.